File: jtms/java/Simplify.java
// Algebraic simplifier.
// Translated from simplify.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.
package jtms;

import java.util.*;

/**
 * Simplify provides algebraic simplification using rule-based rewriting.
 */
public class Simplify {

    private static Map<String, Object> cache = new HashMap<>();

    /** Simplify an expression with caching. Corresponds to simplify. */
    public static Object simplify(Object exp) {
        String key = String.valueOf(exp);
        if (cache.containsKey(key)) return cache.get(key);
        Object result = simplifyIt(exp);
        cache.put(key, result);
        return result;
    }

    /** Clear the simplification cache. */
    public static void clearCache() {
        cache.clear();
    }

    @SuppressWarnings("unchecked")
    private static Object simplifyIt(Object exp) {
        Object simplified;
        if (exp instanceof List) {
            List<Object> list = (List<Object>) exp;
            List<Object> result = new ArrayList<>();
            for (Object e : list) result.add(simplify(e));
            simplified = result;
        } else {
            simplified = exp;
        }
        // In a full implementation, algebra rules would be applied here
        return simplified;
    }

    /** Check if exp1 occurs in exp2. Corresponds to occurs-in?. */
    @SuppressWarnings("unchecked")
    public static boolean occursIn(Object exp1, Object exp2) {
        if (JData.deepEqual(exp1, exp2)) return true;
        if (exp2 == null) return false;
        if (exp2 instanceof List) {
            for (Object e : (List<Object>) exp2) {
                if (occursIn(exp1, e)) return true;
            }
        }
        return false;
    }

    /** Check if expression equals a numeric constant. Corresponds to same-constant?. */
    public static boolean sameConstant(Object exp, double constant) {
        if (!(exp instanceof Number)) return false;
        double val = ((Number) exp).doubleValue();
        return Math.abs(val - constant) < Match.TOLERANCE;
    }

    /** Check if expression is zero. */
    public static boolean isZero(Object exp) { return sameConstant(exp, 0); }

    /** Check if expression is one. */
    public static boolean isOne(Object exp) { return sameConstant(exp, 1); }

    /** Algebraic ordering predicate. Corresponds to alg<. */
    @SuppressWarnings("unchecked")
    public static boolean algLess(Object e1, Object e2) {
        if (Match.matchEqual(e1, e2)) return false;
        boolean e1IsList = e1 instanceof List;
        boolean e2IsList = e2 instanceof List;
        if (e1IsList) {
            if (e2IsList) {
                List<Object> l1 = (List<Object>) e1;
                List<Object> l2 = (List<Object>) e2;
                if (l1.isEmpty() && l2.isEmpty()) return false;
                if (l1.isEmpty()) return true;
                if (l2.isEmpty()) return false;
                if (Match.matchEqual(l1.get(0), l2.get(0))) {
                    return algLess(l1.subList(1, l1.size()), l2.subList(1, l2.size()));
                }
                return algLess(l1.get(0), l2.get(0));
            }
            return false;
        }
        if (e2IsList) return true;
        boolean e1IsStr = e1 instanceof String;
        boolean e2IsStr = e2 instanceof String;
        if (e1IsStr) {
            return e2IsStr && ((String) e1).compareTo((String) e2) < 0;
        }
        if (e2IsStr) return true;
        if (e1 instanceof Number && e2 instanceof Number) {
            return ((Number) e1).doubleValue() < ((Number) e2).doubleValue();
        }
        return false;
    }
}

File: jtms/go/simplify.go
// Algebraic simplifier.
// Translated from simplify.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.

package jtms

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// SimplifyCache caches previously simplified expressions.
var SimplifyCache = make(map[string]interface{})

// AlgebraRules holds the simplification rules.
// Each rule is a slice: [pattern, predicate, skeleton].
var AlgebraRules [][]interface{}

func init() {
	AlgebraRules = buildAlgebraRules()
}

// Simplify simplifies an algebraic expression using cached results.
// Corresponds to the Lisp function simplify.
func Simplify(exp interface{}) interface{} {
	key := fmt.Sprintf("%v", exp)
	if cached, ok := SimplifyCache[key]; ok {
		return cached
	}
	result := simplifyIt(exp, AlgebraRules)
	SimplifyCache[key] = result
	return result
}

// ClearSimplifyCache clears the simplification cache.
func ClearSimplifyCache() {
	SimplifyCache = make(map[string]interface{})
}

// simplifyIt recursively simplifies an expression by applying rules.
func simplifyIt(exp interface{}, rules [][]interface{}) interface{} {
	// First simplify subexpressions
	var simplified interface{}
	if expSlice, ok := exp.([]interface{}); ok {
		result := make([]interface{}, len(expSlice))
		for i, e := range expSlice {
			result[i] = Simplify(e)
		}
		simplified = result
	} else {
		simplified = exp
	}
	result := tryMatcherRules(simplified, rules)
	if deepEqual(result, exp) {
		return result
	}
	return simplifyIt(result, rules)
}

// tryMatcherRules tries to apply simplification rules to an expression.
func tryMatcherRules(exp interface{}, rules [][]interface{}) interface{} {
	for _, rule := range rules {
		if len(rule) < 3 {
			continue
		}
		pattern := rule[0]
		predicate := rule[1]
		skeleton := rule[2]

		bindings, ok := Match(pattern, exp, nil)
		if !ok {
			continue
		}
		if checkPredicate(predicate, bindings) {
			return SubstituteIn(skeleton, bindings)
		}
	}
	return exp
}

// checkPredicate evaluates a predicate with bindings.
func checkPredicate(pred interface{}, bindings Dict) bool {
	if pred == nil {
		return true
	}
	// For Go, predicates are represented as functions
	if fn, ok := pred.(func(Dict) bool); ok {
		return fn(bindings)
	}
	return true
}

// AlgLess is a sort predicate for algebraic expressions.
// Corresponds to the Lisp function alg<.
func AlgLess(e1, e2 interface{}) bool {
	if MatchEqual(e1, e2) {
		return false
	}
	e1s, e1IsList := e1.([]interface{})
	e2s, e2IsList := e2.([]interface{})
	if e1IsList {
		if e2IsList {
			if len(e1s) == 0 && len(e2s) == 0 {
				return false
			}
			if len(e1s) == 0 {
				return true
			}
			if len(e2s) == 0 {
				return false
			}
			if MatchEqual(e1s[0], e2s[0]) {
				return AlgLess(e1s[1:], e2s[1:])
			}
			return AlgLess(e1s[0], e2s[0])
		}
		return false
	}
	if e2IsList {
		return true
	}
	e1Str, e1IsStr := e1.(string)
	e2Str, e2IsStr := e2.(string)
	if e1IsStr {
		if e2IsStr {
			return strings.Compare(e1Str, e2Str) < 0
		}
		return false
	}
	if e2IsStr {
		return true
	}
	e1f, e1IsNum := toFloat(e1)
	e2f, e2IsNum := toFloat(e2)
	if e1IsNum && e2IsNum {
		return e1f < e2f
	}
	return false
}

// SameConstant checks if an expression equals a numeric constant.
func SameConstant(exp interface{}, constant float64) bool {
	f, ok := toFloat(exp)
	if !ok {
		return false
	}
	return math.Abs(f-constant) < MatchTolerance
}

// IsZero checks if an expression is zero.
func IsZero(exp interface{}) bool { return SameConstant(exp, 0) }

// IsOne checks if an expression is one.
func IsOne(exp interface{}) bool { return SameConstant(exp, 1) }

// IsPlusMult checks if exp is + or *.
func IsPlusMult(exp interface{}) bool {
	s, ok := exp.(string)
	return ok && (s == "+" || s == "*")
}

// OccursIn checks if exp1 occurs anywhere in exp2.
func OccursIn(exp1, exp2 interface{}) bool {
	if deepEqual(exp1, exp2) {
		return true
	}
	if exp2 == nil {
		return false
	}
	if s, ok := exp2.([]interface{}); ok {
		for _, e := range s {
			if OccursIn(exp1, e) {
				return true
			}
		}
	}
	return false
}

// IsSorted checks if a list is sorted by the given predicate.
func IsSorted(list []interface{}, less func(a, b interface{}) bool) bool {
	for i := 1; i < len(list); i++ {
		if less(list[i], list[i-1]) {
			return false
		}
	}
	return true
}

// SortAlg sorts a list by the algebraic ordering.
func SortAlg(list []interface{}) []interface{} {
	result := make([]interface{}, len(list))
	copy(result, list)
	sort.SliceStable(result, func(i, j int) bool {
		return AlgLess(result[i], result[j])
	})
	return result
}

// buildAlgebraRules creates the standard algebra simplification rules.
func buildAlgebraRules() [][]interface{} {
	// Simplified representation of algebra rules.
	// In the full Lisp version, these are match patterns with predicates.
	// Here we represent them as Go functions for practicality.
	return [][]interface{}{
		// Rules are applied in order; each is [pattern, predicate, skeleton]
		// For a full port, each pattern/skeleton would use the match.go system.
		// This is a structural placeholder showing the rule architecture.
	}
}
